# Map, Filter, Reduce, Lambda & Recursion

## Tasks Today:

1) <b>Lambda Functions</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Saving to a Variable <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Multiple Inputs <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Passing a Lambda into a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; e) Returning a Lambda from a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; f) In-Class Exercise #1 <br>
2) <b>Map</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Map <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #2 <br>
3) <b>Filter</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Filter <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #3 <br>
4) <b>Reduce</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Reduce <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #4 <br>
5) <b>Recursion</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Implementing a Base <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Writing a Factorial Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #5 <br>
6) <b>Generators & Iterators</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Yield Keyword <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Inifinite Generator <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #6 <br>
7) <b>Exercises</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Exercise #1 - Filtering Empty Strings <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Exercise #2 - Sorting with Last Name <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Exercise #3 - Conversion to Farhenheit <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Exercise #4 - Fibonacci Sequence <br>

## Lambda Functions <br>
<p>Lambda functions... or "Anonymous Functions" are referring to inline functions with no name. The keyword lambda denotes the no name function, and executes within a single line. Without saving it to a variable; however, it is not able to be used, unless passed in either as a paramater or within list comprehension.<br>Written as "(keyword lambda) (one or more inputs) (colon) (function to be executed)"</p>

#### Syntax

In [3]:
# Lambda function is an "anonymous functions. There are ways to give it a "name" but specifically you just use it and it does 
# what it needs to do before moving on. Mostly used for one time use. 

# Lambda is a throwaway function that you create in the moment you need it. Since it takes up less space, the computer 
# Saves it, uses it and then discards it after its done. 

def addTwo(x):
    return x+2

print(addTwo(4))

# Lambda Function Syntax
# Eample of calling a lambda function without a variable

print((lambda x:x+2)(4))

6


#### Saving to a Variable

In [ ]:
f_test = lambda x: x+2
f_test(4)

# difference between using a lambda or a def fuc: lambda is "slightly" smaller. If you have a lot of moving parts.
# it's better to use the function. It really depends on the solution you are looking for. If you need a one liner solution
# lambda function is better to use. 

#### Multiple Inputs

In [5]:
# Multiple Inputs with no variable 
print ((lambda x,y,z: x * y * z)(3,5,8))

# Multiple Inputs with a variable attached 
x_test = lambda x,y,z: x * y * z 
print (x_test(3,5,8))

120
120


#### Passing a Lambda into a Function

In [6]:
def multiply(f, num): 
    """
        f expects a lambda fucntion
        num expects a number 
    """
    return f(num)
multiply(lambda x: x * x, 4) #multiply function has 2 paramaters 'x * x' == is the first
# Lambda 4: 4 * 4

16

#### Returning a Lambda from a Function

In [14]:
# Regular defined function
def multiply_test(num): 
    return num * 4 

# Function within a function
def returnFunc(): 
    test = 4
    def multiply(num):
        return num * 2
#     return multiply, test # creating a tuple. so if I want to get to the mulitply function. 
    return multiply

f_return = returnFunc() # recalling def returnFunc
# print (f_return(4)) # because we returned mutlply here, we can actually...
# print(returnFunc()) 

print(returnFunc()) # when you call on returnFunc. It just tells you what it has inside of it. a multply at a specific area
# If you want access to that. Call it first "f_return". To actually use it "print(f_return)"
print(f_return(4)) 

# Lambda function returned from a regular function 
def returnLam(b,c): 
    return lambda x,a: x + a + b + c
r_lamb = returnLam(4,5)
print (r_lamb(5,5))

<function returnFunc.<locals>.multiply at 0x0000020F06531F70>
8
19


#### If Statements within Lambdas

In [15]:
# Lambda x : True if (condition) else False
f_condition = lambda num : num * 2 if num > 10 else num + 2

print(f_condition(8))
print(f_condition(12))

10
24


#### In-Class Exercise #1 <br>
<p>Write an anonymous function that cubes the arguments passed in and assign the anonymous function to a variable 'f'.</p>

In [16]:
f = lambda cube: cube ** 3

print(f(2))
print(f(3))
print(f(4))

8
27
64


## Map <br>
<p>The map function allows you to iterate over an entire list while running a function on each item of the list. This is why the map function works well with lambda's, because it simplifies things and you write less lines of code.<br>The syntax for a map function is "map(function to be used, list to be used)"<br>However, you must be careful, as the map function returns a map object, not a list. To turn it into a list we use the list() type conversion.</p>

#### Syntax

In [19]:
# map(func,iterable(list,dict,tuple,set))
# Normally the usage of map happens with a pre-defined function -- but we can use lambdas as well
def squared(num, num2):
    if num < 10 and num2 < 10:
        return num ** 2, num2 ** 2
    else: 
        return num, num2

numbers = [4,11,20,3,15]
more_nums = [4,10,3,2,6]

squared_nums_map = list(map(squared, numbers, more_nums)) 
#Wraps map() in a list call. It also changes map(None, x) to list(x). 
#Using from future_builtins import map disables this fixer.
print(squared_nums_map)



[(16, 16), (11, 10), (20, 3), (9, 4), (15, 6)]


#### Using Lambda's with Map

In [23]:
# map(lambda x: x + 2, list)
# using lambda in map happens in line (or in one line) usually
# list(map(lambda x : x * 2,nums))
squared_nums_lamb = list(map(lambda x,y: (x ** 2, y ** 2) if x < 10 and y < 10 else (x,y), numbers, more_nums))
print(squared_nums_lamb)



[(16, 16), (11, 10), (20, 3), (9, 4), (15, 6)]


#### In-Class Exercise #2 <br>
<p>Use the map function to double each number and minus it by one in the list by using a lambda function</p>

In [31]:
double_num = list(map(lambda x,y: (((x * 2) - 1), ((y * 2) - 1)) if x > 0 and y > 0 else (x,y), numbers, more_nums))
print(double_num)
                   

[(7, 7), (21, 19), (39, 5), (5, 3), (29, 11)]


## Filter() <br>
<p>Filter's are similar to the map function, where you're able to pass a function argument and a list argument and filter out something from the list based on the conditions passed. Similar to the map function, it returns a filter object, so you need to type convert it to a list()</p>

#### Syntax

In [33]:
names = ['Bob','Andy','Max','Evan','Angelica']

# what this function looks for is true or false values

def a_names(name): 
    if name[0].lower() == "a":
        return True
    else:
        return False # if you want to filter out names that start with "a" switch True and False
new_names_list = list(filter(a_names,names))
print(new_names_list)

['Bob', 'Max', 'Evan']


#### Using Lambda's with Filter()

In [35]:
new_names_lamb = list(filter(lambda name: True if name[0].lower() == "a" else False, names))
# lambda is taking the place of "a_names"
print (new_names_lamb)

['Andy', 'Angelica']


#### In-Class Exercise #3 <br>
<p>Filter out all the numbers that are below the mean of the list.<br><b>Hint: Import the 'statistics' module</b></p>

In [46]:
from statistics import mean

nums = [2,7,4.2,1.6,9,4.4,4.9]

print(mean(nums)) 

avg = mean(nums) # defined that result of mean(nums) == 4.72

print(avg)

def filterNum(name):
    if name < avg:
        return True
    else: 
        return False

filter_num_list = list(filter(filterNum,nums))
print(filter_num_list)

# Lambda function solution
below_avg = list(filter(lambda x: True if x <= mean(nums) else False, nums))
print (below_avg)

4.728571428571429
4.728571428571429
[2, 4.2, 1.6, 4.4]
[2, 4.2, 1.6, 4.4]


## Reduce() <br>
<p>Be very careful when using this function, as of Python 3 it's been moved to the 'functools' library and no longer is a built-in function.<br>The creator of Python himself, says to just use a for loop instead.</p>

#### Syntax

In [47]:
from functools import reduce

# reduce(function,iterable)

list_1 = [2,4,6,8,10]

def addNums(num1,num2):
    return num1 + num2

result_add = reduce(addNums, list_1)

print(result_add)


30


#### Using Lambda's with Reduce()

#### In-Class Exercise #4 <br>
<p>Use the reduce function to multiply the numbers in the list below together with a lambda function.</p>

In [ ]:
my_list = [1,2,3,4]

from functools import reduce



## Recursion <br>
<p>Recursion means that a function is calling itself, so it contanstly executes until a base case is reached. It will then push the returning values back up the chain until the function is complete. A prime example of recursion is computing factorials... such that 5! (factorial) is 5*4*3*2*1 which equals 120.</p>

#### Implementing a Base Case

In [51]:
def addNums(num):
    # Set base case for recusive function
    if num <= 1: 
        print("addNums(1) = 1")
    else: 
        print(f"addNums({num}) = {num} + addNums({num - 1})")
        return num + addNums(num -1)

#### Writing a Factorial Function

In [52]:
# 5! = 5 * 4 * 3 * 2 * 1

def factorial(num):
    if num <= 1: 
        return 1
    else: 
        return num * factorial(num - 1)

factorial(5) 


120

#### In-Class Exercise #5 <br>
<p>Write a recursive function that subtracts all numbers to the argument given.</p>

In [57]:
# return num - subNums(num - 1)

def subNum(num):
    if num <= 1:
        print ("subNums(1) = 1") 
        return num
    else:
        print (f"subNums({num}) = {num} - subNums({num - 1})")
        return num - subNum(num - 1)

subNum(10)

subNums(10) = 10 - subNums(9)
subNums(9) = 9 - subNums(8)
subNums(8) = 8 - subNums(7)
subNums(7) = 7 - subNums(6)
subNums(6) = 6 - subNums(5)
subNums(5) = 5 - subNums(4)
subNums(4) = 4 - subNums(3)
subNums(3) = 3 - subNums(2)
subNums(2) = 2 - subNums(1)
subNums(1) = 1


5

## Generators <br>
<p>Generators are a type of iterable, like lists or tuples. They do not allow indexing, but they can still be iterated through with for loops. They are created using functions and the yield statement.</p>

#### Yield Keyword <br>
<p>The yield keyword denotes a generator, it doesn't return so it won't leave the function and reset all variables in the function scope, instead it yields the number back to the caller.</p>

In [1]:
def my_range(stop, start, step = 2):
    while start < stop: 
        yield start
        # instead of returning the value of start it will keep going. denotes a generator 
        start += step 
for i in my_range(20, start = 2):
    print(i)

my_range(20,start = 2)

2
4
6
8
10
12
14
16
18


<generator object my_range at 0x000001CA9BB2FF90>

#### Infinite Generator

In [ ]:
# bad, never create infinite loops

#### In-Class Exercise #6 <br>
<p>Create a generator that takes a number argument and yields that number squared, then prints each number squared until zero is reached.</p>

In [ ]:
def squared(stop, start, step = -1):
    while start < stop: 
        yield start
        start += step

for i in squared(20, start = 3):
    if i >= 0: 
        print(i**2)

9
4
1
0


# Exercises

### Exercise #1 <br>
<p>Filter out all of the empty strings from the list below</p>

`Output: ['Argentina', 'San Diego', 'Boston', 'New York']`

In [15]:
places = [" ","Argentina", " ", "San Diego","","  ","","Boston","New York"]
# Use Filter. 
def placesNames(): 
    ' '.join(lstr).split()
    
destination = list(filter(str.strip,places))
print(destination)

['Argentina', 'San Diego', 'Boston', 'New York']


### Exercise #2 <br>
<p>Write an anonymous function that sorts this list by the last name...<br><b>Hint: Use the ".sort()" method and access the key"</b></p>

`Output: ['Victor aNisimov', 'Gary A.J. Bernstein', 'Joel Carter', 'Andrew P. Garfield', 'David hassELHOFF']`

In [14]:
author = ["Joel Carter", "Victor aNisimov", "Andrew P. Garfield","David hassELHOFF","Gary A.J. Bernstein"]

#lambda function. That sorts list by last name. HINT: Use the .sort() and use your lambda function inside of. map is 
# the most adventageous to use. Sort by last name. You want to use split on it. And sort it by the last name. 


author.sort(key=lambda x:x.lower().split()[-1])

print(author)

['Victor aNisimov', 'Gary A.J. Bernstein', 'Joel Carter', 'Andrew P. Garfield', 'David hassELHOFF']


### Exercise #3 <br>
<p>Convert the list below from Celsius to Farhenheit, using the map function with a lambda...</p>

`Output: [('Nashua', 89.6), ('Boston', 53.6), ('Los Angelos', 111.2), ('Miami', 84.2)]

In [32]:
# F = (9/5)*C + 32
places = [('Nashua',32),("Boston",12),("Los Angelos",44),("Miami",29)]

# inside of your lambda function you want to target the numbers. and place them inside of the forumla.
f_list = list(map(lambda x,y: (x, y*(9/5)+ 32), places))
print(places)


TypeError: <lambda>() missing 1 required positional argument: 'y'

### Exercise #4 <br>
<p>Write a recursion function to perform the fibonacci sequence up to the number passed in.</p>

`Output for fib(5) => 
Iteration 0: 1
Iteration 1: 1
Iteration 2: 2
Iteration 3: 3
Iteration 4: 5
Iteration 5: 8`

In [35]:
def fibonacci (num):
    if num <= 1: 
        return 1
    elif num <= 2:
        return 1
    else:
        return num * fibonacci(num - 1)

# every 3rd number is a multiple of 2
# every 4th number is a multiple of 3
# every 5th numbrt id s mulyiplr of 5
fibonacci (5) 

SyntaxError: invalid syntax (<ipython-input-35-1c34b42be3ad>, line 4)

In [ ]:
# only names that start with "aeiou"

names = ['Nate', "andrea", 'Joel', "oliver", 'Gabby', 'Ivan']

def names_vowel(x):
    return x[0].lower() in "aeiou"

fil_names = filter(names_vowel, names)
print(list(fil_names)

In [ ]:
a_list = ['hello', 'I', 'love', 'you']; 

def addWords(accum, nextWord):
    return accum + " " + nextWord

result = reduce(addWords, a_list)
print(result)

